source/common/http/http2/codec_impl.cc

```diff
@@ -210,8 +234,13 @@ void ConnectionImpl::StreamImpl::pendingRecvBufferLowWatermark() {
   readDisable(false);
 }
 
-void ConnectionImpl::ClientStreamImpl::decodeHeaders() {
+void ConnectionImpl::ClientStreamImpl::decodeHeaders(nghttp2_headers_category category) {
   auto& headers = absl::get<ResponseHeaderMapPtr>(headers_or_trailers_);
+  if (category == NGHTTP2_HCAT_RESPONSE &&
+      CodeUtility::is1xx(Http::Utility::getResponseStatus(*headers))) {
+    waiting_for_non_informational_headers_ = true;
```

Comment:
there's more semantic changes than I expected here; I wonder if we could get @alyssar to take a look at these codec changes?

Reply:
Sorry this change was relatively "simple" so I kept it in here. I'm happy to split it out but @alyssawilk should definitely review this part. LMK if you all prefer that I split it out.